### PR TITLE
Add view number to Pre-Processor messages

### PR DESCRIPTION
**Pull request by Efrat1**
Added view number to the following messages:
PreProcessRequestMsg, PreProcessReplyMsg,
PreProcessBatchRequestMsg, PreProcessBatchReplyMsg,

And add the view number to hash.

### CODE REVIEW

File: bftengine/src/preprocessor/PreProcessor.cpp [hunk 1]
```diff
@@ -821,18 +835,28 @@ bool PreProcessor::checkPreProcessReqPrerequisites(SeqNum reqSeqNum,
...
-      valid = false;
-    } else if (!validateMessage(msg.get())) {
-      preProcessorMetrics_.preProcReqInvalid++;
-      valid = false;
+
+  if (batchReq->viewNum() != myReplica_.getCurrentView()) {
+    LOG_INFO(logger(),
+             "Ignore PreProcessBatchRequestMsg as the view is invalid"
+                 << KVLOG(batchReq->senderId(), batchReq->clientId(), batchReq->getCid(), batchReq->viewNum()));
+    valid = false;
```
Comment (yuliasherman) [inline bftengine/src/preprocessor/PreProcessor.cpp:843]: Call 'return false' here.

File: bftengine/src/preprocessor/PreProcessor.cpp [hunk 2]
```diff
@@ -928,9 +952,14 @@ void PreProcessor::onMessage<PreProcessRequestMsg>(PreProcessRequestMsg *message
   const string reqType = msg->reqType() == REQ_TYPE_PRE_PROCESS ? "REQ_TYPE_PRE_PROCESS" : "REQ_TYPE_CANCEL";
   LOG_DEBUG(logger(),
             "Received PreProcessRequestMsg" << KVLOG(
-
                 reqType, msg->reqSeqNum(), msg->getCid(), msg->senderId(), msg->clientId(), msg->reqOffsetInBatch()));
-  handleSinglePreProcessRequestMsg(msg);
+  if (msg->viewNum() != myReplica_.getCurrentView()) {
+    LOG_INFO(logger(),
+             "Ignore PreProcessRequestMsg as the view is invalid" << KVLOG(
+                 msg->reqSeqNum(), msg->senderId(), msg->clientId(), msg->reqOffsetInBatch(), message->viewNum()));
```
Comment (yuliasherman) [inline bftengine/src/preprocessor/PreProcessor.cpp:959]: In order to identify a request, we need to print in each log clientId, seqNum, and cid. Please ann cid.

File: bftengine/src/preprocessor/PreProcessor.cpp [hunk 3]
```diff
@@ -1014,12 +1053,20 @@ bool PreProcessor::checkPreProcessBatchReplyMsgCorrectness(const PreProcessBatch
...
-    if (!checkPreProcessReplyPrerequisites(
-            replyMsg->reqSeqNum(), replyMsg->getCid(), senderId, batchCid, replyMsg->reqOffsetInBatch()) ||
-        !validateMessage(replyMsg.get())) {
-      valid = false;
-      break;
+  if (batchReply->viewNum() != myReplica_.getCurrentView()) {
+    LOG_INFO(
+        logger(),
+        "Ignore PreProcessBatchReplyMsg as the view is invalid" << KVLOG(batchCid, senderId, batchReply->viewNum()));
+    valid = false;
```
Comment (yuliasherman) [inline bftengine/src/preprocessor/PreProcessor.cpp:1060]: Return 'false' here.

File: bftengine/src/preprocessor/messages/PreProcessResultMsg.cpp [hunk 4]
```diff
@@ -55,7 +55,8 @@ PreProcessResultMsg::PreProcessResultMsg(ClientRequestMsgHeader* body) : ClientR
 std::pair<char*, uint32_t> PreProcessResultMsg::getResultSignaturesBuf() { return getExtraBufPtr(); }
 
 std::optional<std::string> PreProcessResultMsg::validatePreProcessResultSignatures(ReplicaId myReplicaId,
-                                                                                   int16_t fVal) {
+                                                                                   int16_t fVal,
```
Comment (yuliasherman) [inline bftengine/src/preprocessor/messages/PreProcessResultMsg.cpp:58]: We decided at the first stage not to touch this message. Please revert all related changes.

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 5]
```diff
@@ -1015,12 +1015,13 @@ bool ReplicaImp::validatePreProcessedResults(const PrePrepareMsg *msg, const Vie
       const MessageBase::Header *hdr = (MessageBase::Header *)requestBody;
       if (hdr->msgType == MsgCode::PreProcessResult) {
         tasks.push_back(threadPool.async(
-            [&errors, requestBody, error_id](auto replicaId, auto fVal) {
+            [&errors, requestBody, error_id](auto replicaId, auto fVal, auto viewNum) {
```
Comment (yuliasherman) [inline bftengine/src/bftengine/ReplicaImp.cpp:1018]: Did you touch PreProcessResultMsg? We decided not to touch it for now. Please revert.

**Review by yuliasherman — CHANGES_REQUESTED** [review]

File: bftengine/src/preprocessor/tests/preprocessor_test.cpp [hunk 6]
```diff
@@ -58,6 +58,7 @@ PreProcessorRecorder preProcessorRecorder;
 std::shared_ptr<concord::performance::PerformanceManager> sdm = make_shared<concord::performance::PerformanceManager>();
 
 uint64_t reqRetryId = 20;
+ViewNum viewNum = 1;
```
Comment (yuliasherman) [inline bftengine/src/preprocessor/tests/preprocessor_test.cpp:61]: As you don't modify this number anywhere => define the const variable in the beginning of the file.

File: bftengine/src/preprocessor/tests/preprocessor_test.cpp [hunk 7]
```diff
@@ -927,13 +931,24 @@ TEST(requestPreprocessingState_test, handlePreProcessBatchRequestMsg) {
...
     batch.push_back(preProcessReqMsg);
     overallReqSize += preProcessReqMsg->size();
   }
-  auto* preProcessBatchReqMsg =
-      new PreProcessBatchRequestMsg(REQ_TYPE_PRE_PROCESS, clientId, 1, batch, cid, overallReqSize);
+  auto* preProcessBatchReqMsg = new PreProcessBatchRequestMsg(
+      REQ_TYPE_PRE_PROCESS, clientId, 1, batch, cid, overallReqSize, replica.getCurrentView());
   auto msgHandlerCallback = msgHandlersRegPtr->getCallback(bftEngine::impl::MsgCode::PreProcessBatchRequest);
   msgHandlerCallback(preProcessBatchReqMsg);
   usleep(waitForExecTimerMillisec * 1000);
```
Comment (yuliasherman) [inline bftengine/src/preprocessor/tests/preprocessor_test.cpp:954]: Add tests for rejected messages.

Comment (Efrat1) [reply bftengine/src/preprocessor/tests/preprocessor_test.cpp:954]: Will add

File: bftengine/src/preprocessor/tests/preprocessor_test.cpp (hunk 6 above)
Comment (Efrat1) [reply bftengine/src/preprocessor/tests/preprocessor_test.cpp:61]: Already done at preprocessor_test.cpp.

**Review by yuliasherman — APPROVED** [review]
Looks great! Thanks.

**Pull request merged**